### PR TITLE
xds: create singleton XdsClient object (promote ClientXdsClient)

### PR DESCRIPTION
**Pull request by voidzcy**
Implemented `SharedXdsClientPoolProvider`, which is a global `XdsClientPoolFactory` that multiple gRPC clients will call to obtain an `ObjectPool<XdsClient>` instance. 

- The first call of `SharedXdsClientPoolProvider.getXdsClientPool()` reads the xDS bootstrap file and creates the xDS channel. `XdsInitializationException`s are propagated up from the downstreams. It returns a thread-safe `ObjectPool<XdsClient>` with ref-counting applied to the XdsClient it returns to its callers.
- The first call of `ObjectPool<XdsClient>.getObject()` creates the XdsClient instance. 

Note, it seems not ideal to read bootstrap file and create the xDS channel before the actually creating the XdsClient instance. But `ObjectPool<XdsClient>.getObject()` interface does not throw any exception, we have no way for error propagation (throwing RuntimeException may not good for its API). This is one limitation of the implementation.

This PR promotes `ClientXdsClient` introduced in #7533, which is an XdsClient implementation with its independent synchronization mechanism.

### CODE REVIEW

File: xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    latch.await(1L, TimeUnit.SECONDS);
+    assertThat(xdsClientRef.get()).isNotNull();
+    for (int i = 0; i < numThreads; i++) {
+      assertThat(xdsClients[i]).isSameInstanceAs(xdsClientRef.get());
+    }
+    executor.shutdown();
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_concurrentReturn() throws InterruptedException {
```
Comment (dapengzhang0) [inline xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:213]: This kind of test runs non-deterministically, not fix for unit tests. I don't think we have to cover concurrency in unit tests.

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:213]: Why is it not deterministic?

Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:213]: The thread pool execute multiple tasks concurrently, the execution order of concurrent runnable instructions is not deterministic.

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:213]: > The thread pool execute multiple tasks concurrently, the execution order of concurrent runnable instructions is not deterministic.

That's indeed the purpose of testing it. It should still work correctly. Without concurrency, the unit tests will be trivial.

Comment (dapengzhang0) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:213]: That's why it's not necessary to test concurrency in unit tests. It won't deterministically catch a bug.

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java]: Alright, deleted those concurrency tests.

**Review by dapengzhang0 — APPROVED** [review]

**Comment by sanjaypujare** [comment]
Is it possible to not use the singleton XdsClient if we run into issues? For example, use an environment variable to switch between the old and new way just in case we have to fall back to the old mech.

**Comment by voidzcy** [comment]
> Is it possible to not use the singleton XdsClient if we run into issues? For example, use an environment variable to switch between the old and new way just in case we have to fall back to the old mech.

It is possible. I think the easiest way to achieve that is to split the code path in the global factory (`SharedXdsClientPoolProvidere`) that creates the XdsClient object pool. That is, the global factory can run in two modes: one for always creating and returning a new `ObjectPool<XdsClient>` and the other for returning the cached object pool if it has already been created before.

But all of our existing use cases are assuming there is only one client channel, so there is no difference between the XdsClient is for the channel or for the process. I don't think running multiple channels with each channel using its own XdsClient makes sense, and that's the reason we are implementing shared XdsClient between channels.

**Comment by sanjaypujare** [comment]
> > Is it possible to not use the singleton XdsClient if we run into issues? For example, use an environment variable to switch between the old and new way just in case we have to fall back to the old mech.
> 
> It is possible. I think the easiest way to achieve that is to split the code path in the global factory (`SharedXdsClientPoolProvidere`) that creates the XdsClient object pool. That is, the global factory can run in two modes: one for always creating and returning a new `ObjectPool<XdsClient>` and the other for returning the cached object pool if it has already been created before.
> 
> But all of our existing use cases are assuming there is only one client channel, so there is no difference between the XdsClient is for the channel or for the process. I don't think running multiple channels with each channel using its own XdsClient makes sense, and that's the reason we are implementing shared XdsClient between channels.

Not debating what's the right approach. I am talking about implementation bugs or quirks in this code biting us at the wrong time (say in production) in which case I just want to fall back on the old implementation since that has been working for us so far.

What would it take to add this conditional/env var support in this PR?

**Comment by voidzcy** [comment]
> Not debating what's the right approach. I am talking about implementation bugs or quirks in this code biting us at the wrong time (say in production) in which case I just want to fall back on the old implementation since that has been working for us so far.
> 
> What would it take to add this conditional/env var support in this PR?

If you have only one channel, there is no difference in creating the XdsClient instance by itself or by using the factory. The creation happens exactly once. The behavior is identical, it's only about where the code is located.

**Comment by sanjaypujare** [comment]
> 
> If you have only one channel, there is no difference in creating the XdsClient instance by itself or by using the factory. The creation happens exactly once. The behavior is identical, it's only about where the code is located.

And if there is more than one channel? Is there a possibility that a user might run into an issue?

**Comment by voidzcy** [comment]
> And if there is more than one channel? Is there a possibility that a user might run into an issue?

We are implementing to support multiple channels by using a shared XdsClient. This is the purpose of this change. Having multiple XdsClient per process is non-goal.

**Pull request merged**